presto-test-jdbc-compatibility-old-server/src/test/java/io/prestosql/TestJdbcResultSetCompatibilityOldServer.java

```diff
@@ -0,0 +1,121 @@
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql;
+
+import com.google.common.collect.ImmutableList;
+import io.prestosql.jdbc.BaseTestJdbcResultSet;
+import io.prestosql.jdbc.PrestoDriver;
+import io.prestosql.testing.TestngUtils;
+import org.testcontainers.containers.PrestoContainer;
+import org.testng.annotations.AfterClass;
+import org.testng.annotations.BeforeClass;
+import org.testng.annotations.DataProvider;
+import org.testng.annotations.Factory;
+import org.testng.annotations.Test;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.util.Optional;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Throwables.getStackTraceAsString;
+import static java.util.Objects.requireNonNull;
+
+@Test(singleThreaded = true)
+public class TestJdbcResultSetCompatibilityOldServer
+        extends BaseTestJdbcResultSet
+{
+    private static final int NUMBER_OF_TESTED_VERSIONS = 5;
+    private static final int TESTED_VERSIONS_GRANULARITY = 3;
+
+    /**
+     * Empty means that we could not obtain current Presto version and tests defined here will be marked as failed.
+     */
+    private final Optional<String> testedPrestoVersion;
+    private PrestoContainer<?> prestoContainer;
+
+    @Factory(dataProvider = "testedPrestoVersions")
+    public TestJdbcResultSetCompatibilityOldServer(Optional<String> testedPrestoVersion)
+    {
+        this.testedPrestoVersion = requireNonNull(testedPrestoVersion, "testedPrestoVersion is null");
+    }
+
+    @DataProvider
+    public static Object[][] testedPrestoVersions()
+    {
+        try {
+            String currentVersionString = requireNonNull(PrestoDriver.class.getPackage().getImplementationVersion());
+            // this should match git-describe output which is used as version stored in metadata
+            Matcher matcher = Pattern.compile("(\\d+)(?:-\\d+-[a-z0-9]+)?").matcher(currentVersionString);
+            checkState(matcher.matches());
+            int currentVersion = Integer.parseInt(matcher.group(1));
+            ImmutableList.Builder<String> testedPrestoVersions = ImmutableList.builder();
+            int lastReleasedVersion = currentVersion - 1;
+            for (int i = 0; i < NUMBER_OF_TESTED_VERSIONS; i++) {
+                testedPrestoVersions.add(String.valueOf(lastReleasedVersion - TESTED_VERSIONS_GRANULARITY * i));
```

Comment:
obviously, this means

- when current is 341-SNAPSHOT, we test aginst 340, 337, ...
- when we do the 341 release, we will be testing against 341, 338

Now, what if `338` does not pass the test for some reason?
The release procedure may result in the master branch turning red.

hm... maybe TESTED_VERSIONS_GRANULARITY is not so good an idea after all?
or maybe the above is actually OK?

Reply:
> obviously, this means
> 
> * when current is 341-SNAPSHOT, we test aginst 340, 337, ...
> * when we do the 341 release, we will be testing against 341, 338

Hmm why so? 340, 337, 334 in both cases. Am I missing sth?

Reply:
sorry, wasn;'t clear. Right after the release, we switch to 342-SNAPSHOT, and then we pick _different_ versions

Reply:
Yeah. It looks like (highly unlikely) we can not notice that given released version is not really compatible with one of the past jdbc drivers.
Maybe if release is non-snapshot we should use GRANRALITY=1 and test more past versions? 
Or aren't we ever running a PR automation when version is non snapshot?

I am fine with dropping granularity here altogether, but it will make tests slightly longer as we probably want to cover servers released in a year or so. Which translates to 10-20 versions.

Reply:
It seems that to trigger issue you mention above we would need to have situation that all below are true:
- driver N is incompatible with server N-2
- driver N-1 compatible with server N-2
- driver N is compatible with server N-1

so we need non-transitive compatibility. Possible but, I would not expect it happen in reality.

Reply:
>so we need non-transitive compatibility. Possible but, I would not expect it happen in reality.

or if/exclusion in the test that makes the test pass (not fail) for a specific version

OK, we can keep as is.